skiko/src/iosMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.ios.kt

```diff
@@ -1,46 +1,64 @@
 package org.jetbrains.skiko.redrawer
 
-import kotlinx.cinterop.addressOf
-import kotlinx.cinterop.autoreleasepool
-import kotlinx.cinterop.objcPtr
-import kotlinx.cinterop.usePinned
-import kotlinx.cinterop.useContents
+import kotlinx.cinterop.*
 import org.jetbrains.skia.BackendRenderTarget
 import org.jetbrains.skia.DirectContext
-import org.jetbrains.skiko.*
+import org.jetbrains.skiko.SkiaLayer
 import org.jetbrains.skiko.context.ContextHandler
 import org.jetbrains.skiko.context.MetalContextHandler
 import platform.CoreGraphics.CGColorCreate
 import platform.CoreGraphics.CGColorSpaceCreateDeviceRGB
 import platform.CoreGraphics.CGContextRef
+import platform.CoreGraphics.CGSizeMake
+import platform.Foundation.NSRunLoop
+import platform.Foundation.NSSelectorFromString
 import platform.Metal.MTLCreateSystemDefaultDevice
 import platform.Metal.MTLDeviceProtocol
 import platform.Metal.MTLPixelFormatBGRA8Unorm
+import platform.QuartzCore.CADisplayLink
 import platform.QuartzCore.CAMetalDrawableProtocol
 import platform.QuartzCore.CAMetalLayer
 import platform.QuartzCore.kCAGravityTopLeft
-import kotlin.system.getTimeNanos
-import platform.CoreGraphics.CGSizeMake
+import platform.darwin.NSObject
 
 internal class MetalRedrawer(
     private val layer: SkiaLayer
 ) : Redrawer {
-    private val contextHandler = MetalContextHandler(layer)
+    internal val contextHandler = MetalContextHandler(layer)
     override val renderInfo: String get() = contextHandler.rendererInfo()
 
     private var isDisposed = false
     internal val device = MTLCreateSystemDefaultDevice()!!
     private val queue = device.newCommandQueue()!!
     private var currentDrawable: CAMetalDrawableProtocol? = null
     private val metalLayer = MetalLayer()
+    private val activeFrameSubscription: Boolean = false
+    private val frameListener = object : NSObject() {
+        @ObjCAction
+        fun onDisplayLinkTick() {
+            if (layer.isShowing()) {
+                draw()
+            }
+        }
+    }
+    private val caDisplayLink = CADisplayLink.displayLinkWithTarget(
+        target = frameListener,
+        selector = NSSelectorFromString(frameListener::onDisplayLinkTick.name)
+    )
 
     init {
         metalLayer.init(this.layer, contextHandler, device)
     }
 
-    private val frameDispatcher = FrameDispatcher(SkikoDispatchers.Main) {
-        if (layer.isShowing()) {
-            draw()
+    private inline fun addFrameSubscription() {
+        if (!activeFrameSubscription) {
+            caDisplayLink.addToRunLoop(NSRunLoop.mainRunLoop, NSRunLoop.mainRunLoop.currentMode)
+        }
+    }
+
+    private inline fun removeFrameSubscription() {
+        if (activeFrameSubscription) {
```

Comment:
It looks like it's always `false` (no assignments to `activeFrameSubscription` at all)

Reply:
Thanks, fixed!